enterprise/server/remote_execution/containers/ociruntime/ociruntime.go

```diff
@@ -773,9 +783,69 @@ func layerPath(layersDir string, hash ctr.Hash) string {
 	return filepath.Join(layersDir, hash.Algorithm, hash.Hex)
 }
 
-// pull downloads and extracts image layers to a directory.
+// ImageStore handles image layer storage for OCI containers.
+type ImageStore struct {
+	layersDir string
+	pullGroup singleflight.Group
+
+	mu           sync.RWMutex
+	cachedLayers map[string][]ctr.Layer
+}
+
+func NewImageStore(layersDir string) *ImageStore {
+	return &ImageStore{
+		layersDir:    layersDir,
+		cachedLayers: map[string][]ctr.Layer{},
+	}
+}
+
+// Pull downloads and extracts image layers to a directory, skipping layers
+// that have already been downloaded, and deduping concurrent downloads for the
+// same layer.
+// Pull always re-authenticates the credentials with the image registry.
 // Each layer is extracted to a subdirectory given by {algorithm}/{hash}, e.g.
 // "sha256/abc123".
+func (s *ImageStore) Pull(ctx context.Context, imageName string, creds oci.Credentials) ([]ctr.Layer, error) {
+	key := hash.Strings(imageName, creds.Username, creds.Password)
+	ch := s.pullGroup.DoChan(key, func() (any, error) {
+		// Use a background ctx to prevent ctx cancellation from one pull
+		// operation affecting all members of this singleflight group.
+		// TODO: use something like github.com/janos/singleflight to ensure we
+		// cancel the download if all group member contexts are cancelled.
+		ctx := context.WithoutCancel(ctx)
+		layers, err := pull(ctx, s.layersDir, imageName, creds)
+		if err != nil {
+			return nil, err
+		}
+
+		s.mu.Lock()
+		s.cachedLayers[imageName] = layers
+		s.mu.Unlock()
+
+		return layers, nil
+	})
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case res := <-ch:
+		if res.Err != nil {
+			return nil, res.Err
+		}
+		return res.Val.([]ctr.Layer), nil
+	}
+}
+
+// CachedLayers returns references to the cached image layers if the image
+// has been pulled. The second return value indicates whether the image has
+// been pulled - if false, the returned slice of layers will be nil.
+func (s *ImageStore) CachedLayers(imageName string) (layers []ctr.Layer, ok bool) {
```

Comment:
looks like the first return value is unused

Reply:
I wound up using it in https://github.com/buildbuddy-io/buildbuddy/pull/6896 which is based on this one - would prefer to keep it in this PR if you don't mind, to avoid having to restack the PRs